### PR TITLE
[MERGE][IMP] web: improve properties selection options ordering

### PR DESCRIPTION
**Pull request by std-odoo**
This merge commits contains 2 improvements aiming to simplify the ordering of
selection options for property fields:

- Correctly insert the new element at the right position when pressing enter
- Allow to drag & drop property values

See underlying commits for details.

Task-3338157

### CODE REVIEW

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 1]
```diff
@@ -198,6 +229,25 @@ export class PropertyDefinitionSelection extends Component {
...
+    /**
+     * Move an option after an other one.
+     *
+     * @param {string} from, the option to move
+     * @param {string} to, the target option
+     *      (null if we move the option at the first index)
+     */
+    onOptionMoveTo(from, to) {
+        const options = this.optionsVisible;
+        const fromIndex = options.findIndex((option) => option[0] == from);
```
Comment (schoffeniels) [inline addons/web/static/src/views/fields/properties/property_definition_selection.js:242]: `===` instead of `==` ?

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 2]
```diff
@@ -198,6 +229,25 @@ export class PropertyDefinitionSelection extends Component {
...
+        let toIndex = options.findIndex((option) => option[0] == to);
+        if (toIndex < fromIndex) {
+            // the first splice operation will change the index
+            toIndex++;
+        }
+        options.splice(toIndex, 0, options.splice(fromIndex, 1)[0]);
+        this.props.onOptionsChange(options);
```
Comment (schoffeniels) [inline addons/web/static/src/views/fields/properties/property_definition_selection.js:249]: Shouldn't we handle the case where we are moving an option that we edited but did not "blur" yet (thus for which the changes were not propagated) ?
For the moment it creates a duplicate key if you move a new option with a value, or it rollbacks to the previous value if you move an existing property that you edited.
Maybe by forcing a blur in `onWillStartDrag`? (not sure if it will work, could maybe mess with the draggable hook)
This could maybe also allow to get rid of the `element.blur()` in the `enable` handler (it seems strange to me to actually do something in it, imo it should only indicate if the Sortable hook is enabled or not, and not depend on the element we try to move).

File: addons/web/static/src/views/fields/properties/property_definition_selection.xml [hunk 3]
```diff
@@ -6,7 +6,11 @@
             <div class="w-100">
                 <div
                     t-foreach="optionsVisible" t-as="option" t-key="option[0]"
-                    class="o_field_property_selection_option d-flex flex-row align-items-center justify-content-between mb-2 rounded">
+                    class="o_field_property_selection_option d-flex flex-row align-items-center justify-content-between mb-2 rounded"
+                    t-att-option="option[0]">
+                    <i t-if="!props.readonly and props.canChangeDefinition"
+                        class="o_field_property_selection_drag oi oi-draggable ui-sortable-handle me-2"
+                        t-att-class="{'invisible': state.emptyOptionIndex === option_index}"/>
```
Comment (schoffeniels) [inline addons/web/static/src/views/fields/properties/property_definition_selection.xml:13]: Shouldn't it be `newOptionIndex` instead of `emptyOptionIndex` ?

**Review by schoffeniels — COMMENTED** [review]
Cool imp for properties again, nice job :)
I left a few comments, do not hesitate to ask if anything is not clear

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 4]
```diff
@@ -12,18 +12,24 @@ export class PropertyDefinitionSelection extends Component {
...
+            if (this.state.newOptionIndex < 0) {
+                return;
+            }
             const inputs = this.propertyDefinitionSelectionRef.el.querySelectorAll(
                 ".o_field_property_selection_option input"
             );
-            if (inputs && inputs.length && !inputs[inputs.length - 1].value) {
-                inputs[inputs.length - 1].focus();
+            if (inputs && inputs.length && !inputs[this.state.newOptionIndex].value) {
+                inputs[this.state.newOptionIndex].focus();
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/property_definition_selection.js:33]: hum, but so we will duplicate the 2 inputs and all their attributes in the template no ? :/

**Review by std-odoo — COMMENTED** [review]
@schoffeniels Hi :) 

Thanks for the validation, I did the changes and answered

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 5]
```diff
@@ -12,20 +13,55 @@ export class PropertyDefinitionSelection extends Component {
...
+                }
+
+                const element = document.activeElement;
+                if (
+                    element &&
+                    element.tagName === "INPUT" &&
+                    element.parentNode.classList.contains("o_field_property_selection_option") &&
+                    (!element.value || !element.value.length)
+                ) {
+                    // try to move an empty option that will be removed
```
Comment (schoffeniels) [inline addons/web/static/src/views/fields/properties/property_definition_selection.js:50]: Not sure what should be the expected behaviour here.
The comment suggests that we cannot move an empty option, but by looking at the condition above I would say that it makes it impossible to move any option if an empty option is focused (could not check on runbot it seems to be dead :/).

If it should prevent to move any option, so that the empty option is first removed, we could maybe use the `.o_field_property_selection_option input:placeholder-shown` selector to simplify the above condition ?

**Review by schoffeniels — COMMENTED** [review]
Hello @std-odoo,
Thanks for your answers :)
I left a few other questions

File: addons/web/static/src/views/fields/properties/property_definition_selection.js (hunk 5 above)
Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/property_definition_selection.js:50]: > but by looking at the condition above I would say that it makes it impossible to move any option if an empty option is focused

Yes that's true, we loose the focus so the option is removed

> If it should prevent to move any option, so that the empty option is first removed, we could maybe use the .o_field_property_selection_option input:placeholder-shown selector to simplify the above condition ?

Sorry, I don't see how it can help, this trick is just to give the autofocus to the new option no ?

**Review by std-odoo — COMMENTED** [review]
@schoffeniels Hi :) 

I answered the questions

Cheers

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 6]
```diff
@@ -12,18 +12,24 @@ export class PropertyDefinitionSelection extends Component {
         // when we create a new option, it's added in the state
         // when we have finished to edit it (blur / enter) we propagate
         // the new value in the props
-        this.state = useState({ newOption: null });
+        this.state = useState({
+            newOption: null,
+            newOptionIndex: -1,
+        });
```
[the remark this answers is not in the document]
Comment (schoffeniels) [reply addons/web/static/src/views/fields/properties/property_definition_selection.js:19]: I wouldn't have put it in the state because imo we should not rerender the component if its value changes (for example in onOptionChange, we may update it using event.value, which triggers a render (which will not change anything), before anyways calling props.onOptionsChange which will also trigger a render.

I guess it was just a personal preference of me here, so let's keep it like this.

**Review by schoffeniels — COMMENTED** [review]
Hello, thanks for the answers :)
I replied but no need to do the changes it's fine like this.
Except maybe for a comment that I believe is misleading ?

File: addons/web/static/src/views/fields/properties/property_definition_selection.js (hunk 5 above)
Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/property_definition_selection.js]: ah ok i see :) yes let's keep it a bit more verbose / explicit

I updated the comment

**Review by std-odoo — COMMENTED** [review]
@schoffeniels Done :)

**Comment by schoffeniels** [comment]
Thanks for the quick change 🏎️
Greened

File: addons/web/static/tests/views/fields/properties_field_tests.js [hunk 7]
```diff
@@ -565,12 +565,35 @@ QUnit.module("Fields", (hooks) => {
...
         assert.strictEqual(options.length, 3, "Should have removed the second option");
-        const optionValues = [...options].map((option) => option.querySelector("input").value);
+        let optionValues = [...options].map((option) => option.querySelector("input").value);
         assert.deepEqual(
             optionValues,
             ["A", "C", "New option"],
             "Should have removed the second option"
         );
+
+        // focus should be in option C
```
Comment (awa-odoo) [inline addons/web/static/tests/views/fields/properties_field_tests.js:579]: Would add a small explanation before this.

Something like:
"Test that pressing 'Enter' inserts a new option after the one currently focused (and not last)."

File: addons/web/static/tests/views/fields/properties_field_tests.js [hunk 8]
```diff
@@ -565,12 +565,35 @@ QUnit.module("Fields", (hooks) => {
...
+            ".o_field_property_selection_option:nth-child(3) input",
+            "New option 2"
+        );
+        options = popover.querySelectorAll(".o_field_property_selection_option");
+        assert.strictEqual(options.length, 4, "Should have added a new option");
+        optionValues = [...options].map((option) => option.querySelector("input").value);
+        assert.deepEqual(
+            optionValues,
+            ["A", "C", "New option 2", "New option"],
+            "Should have added a new option"
```
Comment (awa-odoo) [inline addons/web/static/tests/views/fields/properties_field_tests.js:596]: "... at the correct spot"

File: addons/web/static/tests/views/fields/properties_field_tests.js [hunk 9]
```diff
@@ -565,12 +565,35 @@ QUnit.module("Fields", (hooks) => {
...
+            { key: "Enter" }
+        );
+        await editInput(
+            target,
+            ".o_field_property_selection_option:nth-child(3) input",
+            "New option 2"
+        );
+        options = popover.querySelectorAll(".o_field_property_selection_option");
+        assert.strictEqual(options.length, 4, "Should have added a new option");
+        optionValues = [...options].map((option) => option.querySelector("input").value);
```
Comment (awa-odoo) [inline addons/web/static/tests/views/fields/properties_field_tests.js]: Not sure we need a variable after all, you could just inline this inside deepEqual (same for above).

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 10]
```diff
@@ -32,6 +33,36 @@ export class PropertyDefinitionSelection extends Component {
...
             }
         });
+
+        useSortable({
+            enable: () => {
+                if (!this.props.canChangeDefinition || this.props.readonly) {
+                    return false;
+                }
+
+                const element = document.activeElement;
```
Comment (awa-odoo) [inline addons/web/static/src/views/fields/properties/property_definition_selection.js:43]: "element" -> "focusedElement" ?

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 11]
```diff
@@ -32,6 +33,36 @@ export class PropertyDefinitionSelection extends Component {
...
+                if (
+                    element &&
+                    element.tagName === "INPUT" &&
+                    element.parentNode.classList.contains("o_field_property_selection_option") &&
+                    (!element.value || !element.value.length)
```
Comment (awa-odoo) [inline addons/web/static/src/views/fields/properties/property_definition_selection.js:48]: Testing length === 0 is usually preferred as more explicit.

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 12]
```diff
@@ -32,6 +33,36 @@ export class PropertyDefinitionSelection extends Component {
...
+                ) {
+                    // an empty option has the focus, and so it will be removed when
+                    // we will start moving, so we block the drag and drop in that case
+                    return false;
```
Comment (awa-odoo) [inline addons/web/static/src/views/fields/properties/property_definition_selection.js:52]: I find it a bit sad and not very user friendly to now allow to move a newly created option.

Knowing that you can focus some option and press enter to place it where you want is a bit niche and "power user".
I think most users will naturally press "Add option", then type their text and move it.
But there is a good chance that they also move their option before typing (some people really like their chaussettes sorted :disguised_face: ), in which case it's frustrating as the option disappears.

I understand that this is due to the "blur" mechanism, which is triggered upon clicking on the sorting control, not much you can do.
However, we could change the blur into a global DOM click event, that checks that you clicked outside of the input AND not on the drag&drop element before removing the option.

A bit annoying to do, but if not too complicated I think it would be nice, maybe to discuss with Luc?

If unclear, don't hesitate to ping me :slightly_smiling_face:

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 13]
```diff
@@ -32,6 +33,36 @@ export class PropertyDefinitionSelection extends Component {
...
+                }
+                return true;
+            },
+            ref: this.propertyDefinitionSelectionRef,
+            handle: ".o_field_property_selection_drag",
+            elements: ".o_field_property_selection_option",
+            cursor: "grabbing",
+            onDrop: async ({ element, previous }) => {
+                const from = element.getAttribute("option");
+                const to = previous && previous.getAttribute("option");
```
Comment (awa-odoo) [inline addons/web/static/src/views/fields/properties/property_definition_selection.js:62]: Would renamed "from" and "to" to "movedOption" and "destinationOption".

"from" and "to" would indicate indexes, here from what I understand "from" is the option being moved and "to" is the option after which we want it to go.

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 14]
```diff
@@ -198,6 +229,36 @@ export class PropertyDefinitionSelection extends Component {
...
+     * Move an option after an other one.
+     *
+     * @param {string} from, the option to move
+     * @param {string} to, the target option
+     *      (null if we move the option at the first index)
+     */
+    onOptionMoveTo(from, to) {
+        const options = this.optionsVisible;
+        const fromIndex = options.findIndex((option) => option[0] == from);
+        let toIndex = options.findIndex((option) => option[0] == to);
```
Comment (awa-odoo) [inline addons/web/static/src/views/fields/properties/property_definition_selection.js:243]: Would add a small comment mentioning that if to is null, we get -1, which is intended.

(Had to read the doc of findIndex to understand ^^)

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 15]
```diff
@@ -198,6 +229,36 @@ export class PropertyDefinitionSelection extends Component {
...
+        if (toIndex < fromIndex) {
+            // the first splice operation will change the index
```
Comment (awa-odoo) [inline addons/web/static/src/views/fields/properties/property_definition_selection.js:245]: Wouldn't it be easier to remove the element from the array first, then find the index of the "destinationOption" inside this new array, then insert it right after that?

Would remove the need for this part (I think), which is a bit hard to follow.

File: addons/web/static/src/views/fields/properties/property_definition_selection.scss [hunk 16]
```diff
@@ -3,3 +3,12 @@
         color: $o-main-favorite-color;
     }
 }
+
+.o_field_property_selection_drag {
+    @include o-grab-cursor;
+    color: #adb5bd;
+    text-align: center;
+    &:hover {
+        color: #666666;
```
Comment (awa-odoo) [inline addons/web/static/src/views/fields/properties/property_definition_selection.scss:12]: Do we have to hardcode colors? makes it not super visible in dark mode.

File: addons/web/static/src/views/fields/properties/property_definition_selection.scss [hunk 17]
```diff
@@ -3,3 +3,12 @@
         color: $o-main-favorite-color;
     }
 }
+
+.o_field_property_selection_drag {
+    @include o-grab-cursor;
+    color: #adb5bd;
+    text-align: center;
```
Comment (awa-odoo) [inline addons/web/static/src/views/fields/properties/property_definition_selection.scss:10]: Use bootstrap classes?

File: addons/web/static/src/views/fields/properties/property_definition_selection.xml [hunk 18]
```diff
@@ -6,7 +6,10 @@
             <div class="w-100">
                 <div
                     t-foreach="optionsVisible" t-as="option" t-key="option[0]"
-                    class="o_field_property_selection_option d-flex flex-row align-items-center justify-content-between mb-2 rounded">
+                    class="o_field_property_selection_option d-flex flex-row align-items-center justify-content-between mb-2 rounded"
+                    t-att-option="option[0]">
```
Comment (awa-odoo) [inline addons/web/static/src/views/fields/properties/property_definition_selection.xml:10]: Maybe "option-value" or "option-id" and not just "option" ?

File: addons/web/static/tests/views/fields/properties_field_tests.js [hunk 19]
```diff
@@ -490,9 +491,16 @@ QUnit.module("Fields", (hooks) => {
         const type = popover.querySelector(".o_field_property_definition_type input");
         assert.strictEqual(type.value, "Selection");
 
+        const getOptions = () => {
```
Comment (awa-odoo) [inline addons/web/static/tests/views/fields/properties_field_tests.js:494]: Well I see you changed this here, you may ignore a previous comment then ^^

File: addons/web/static/tests/views/fields/properties_field_tests.js [hunk 20]
```diff
@@ -586,14 +590,33 @@ QUnit.module("Fields", (hooks) => {
...
-        optionValues = [...options].map((option) => option.querySelector("input").value);
         assert.deepEqual(
-            optionValues,
+            getOptionsValues(),
             ["A", "C", "New option 2", "New option"],
             "Should have added a new option"
         );
+
+        await nextTick();
+        const getOptionsElement = (index) => {
```
Comment (awa-odoo) [inline addons/web/static/tests/views/fields/properties_field_tests.js:600]: "getOptionDraggableElement" ?

File: addons/web/static/tests/views/fields/properties_field_tests.js [hunk 21]
```diff
@@ -586,14 +590,33 @@ QUnit.module("Fields", (hooks) => {
...
+                `.o_field_property_selection_option:nth-child(${index + 1})` +
+                    " .o_field_property_selection_drag"
+            );
+        };
+
+        await dragAndDrop(getOptionsElement(0), getOptionsElement(2));
+        assert.deepEqual(
+            getOptionsValues(),
+            ["C", "New option 2", "A", "New option"],
+            "Should have added a new option"
```
Comment (awa-odoo) [inline addons/web/static/tests/views/fields/properties_field_tests.js:611]: Copy/pasta

File: addons/web/static/tests/views/fields/properties_field_tests.js [hunk 22]
```diff
@@ -586,14 +590,33 @@ QUnit.module("Fields", (hooks) => {
...
+        );
+
+        await dragAndDrop(getOptionsElement(3), getOptionsElement(0));
+        assert.deepEqual(
+            getOptionsValues(),
+            ["New option", "C", "New option 2", "A"],
+            "Should have added a new option"
```
Comment (awa-odoo) [inline addons/web/static/tests/views/fields/properties_field_tests.js:618]: Copy/pasta

**Review by awa-odoo — COMMENTED** [review]
Hello @std-odoo 

Neat little feature!
Added some comments and questions to try to polish and maybe make it a bit more user friendly.

Thanks & cheers!

File: addons/web/static/src/views/fields/properties/property_definition_selection.xml (hunk 18 above)
Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/property_definition_selection.xml]: ok for `option-name`

File: addons/web/static/src/views/fields/properties/property_definition_selection.js (hunk 14 above)
Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/property_definition_selection.js]: indeed, not very obvious :)

File: addons/web/static/src/views/fields/properties/property_definition_selection.js (hunk 15 above)
Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/property_definition_selection.js:246]: the code will be less readable imo, we will need to 
- find the option in the list
- remove the option from the list


instead of just fixing the index

File: addons/web/static/src/views/fields/properties/property_definition_selection.js (hunk 12 above)
Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/property_definition_selection.js:51]: > I find it a bit sad and not very user friendly to now allow to move a newly created option.

You can move a new option, if it's not empty (you need to edit it first)

See: https://drive.google.com/file/d/1wpP-Zeuv7FQsBfwsRpIAJ7_y29ewiVQ5/view?usp=sharing

> I think most users will naturally press "Add option", then type their text and move it.

So it will work, as in the video

> However, we could change the blur into a global DOM click event, that checks that you clicked outside of the input AND not on the drag&drop element before removing the option.

But if you start drag & drop an empty option, it won't be removed, but then you cancel the drag & drop (right click e.g.), are you are in a weird state with an empty option, so we also need to add a global event listener for "mouse up", etc 


So the only use case is: you click "Add an option" and you want to move the empty input, but not sure it's worth it to support that, considering the technical cost

**Review by std-odoo — COMMENTED** [review]
@awa-odoo Hi :) 

I did the change / answered

About the "blur" stuff, it work if you edit the option first (which what's most people will do, because the input is focused), so it's only if you want to move an empty option (more info in my comment below)

File: addons/web/static/src/views/fields/properties/property_definition_selection.js (hunk 12 above)
Comment (awa-odoo) [reply addons/web/static/src/views/fields/properties/property_definition_selection.js:51]: Yes, I understand that it works when you type before moving, sorry if it wasn't clear in my message.

I just wanted to see if we could handle the case: I add and I try moving it to the right spot before typing.
But if you think it's niche and not worth the cost, I'm ok with that :slightly_smiling_face:

File: addons/web/static/src/views/fields/properties/property_definition_selection.js (hunk 15 above)
Comment (awa-odoo) [reply addons/web/static/src/views/fields/properties/property_definition_selection.js:246]: Ok for not changing, but can you add an example in this "if" then?

Something like "e.g: I want to move the third option to the second spot, the index I found is X, I need to correct it to Y because when moving it later with the splice operation, it will need to be ..."

(Could also be in the docstring, but then we would need an example that does not go through this part as well ^^)

**Review by awa-odoo — COMMENTED** [review]
Hello @std-odoo 

One last picky comment and we should be good to go :rocket: 

Thanks & cheers!

**Review by std-odoo — COMMENTED** [review]
@awa-odoo Hi :) 

I did the changes, I found a "not that complex" way to support moving an empty option, I created an attribute `_ignoreBlur` to skip the next  `onOptionBlur`, and instead of trigerring a `onOptionChange`, I just fix the index in the state (so if you move an empty option, then refresh the page, you won't get a weird state with an empty option saved)

I did that in a separate commit to ease the review, let me know what you think

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 23]
```diff
@@ -250,16 +238,39 @@ export class PropertyDefinitionSelection extends Component {
...
+        if (
+            activeEl &&
+            this.propertyDefinitionSelectionRef.el.contains(activeEl) &&
+            activeEl.tagName === "INPUT"
+        ) {
+            const optionName = activeEl
+                .closest(".o_field_property_selection_option")
+                .getAttribute("option-name");
+            const editedOptionIndex = options.findIndex((option) => option[0] === optionName);
+            // we might be editing the value and drag and drop something else just after
```
Comment (awa-odoo) [inline addons/web/static/src/views/fields/properties/property_definition_selection.js:251]: If I understood correctly, this is more:

"As we cancel the blur event (to allow moving an empty option without deleting it), we need to manually update the value inside the options to save what the user may have input into it right before moving it around".

File: addons/web/static/src/views/fields/properties/property_definition_selection.js [hunk 24]
```diff
@@ -153,6 +136,9 @@ export class PropertyDefinitionSelection extends Component {
         if (event.target.value && event.target.value.length) {
             // losing the focus on an non-empty option should have no effect
             return;
+        } else if (this._ignoreBlur) {
```
Comment (awa-odoo) [inline addons/web/static/src/views/fields/properties/property_definition_selection.js:139]: Would explain in the docstring what "_ignoreBlur" is used for.

**Review by awa-odoo — COMMENTED** [review]
Hello @std-odoo 

Ok for this implementation, I added a couple comments so that it clearly explains the purpose in the code.

Should we ask for a testing round before merging or is that not necessary?

Thank you for looking into it!

File: addons/web/static/src/views/fields/properties/property_definition_selection.js (hunk 23 above)
Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/property_definition_selection.js:255]: noy only, you can also edit a "non new option" and then move an other one

without this line, the edition won't be saved

**Review by std-odoo — COMMENTED** [review]
@awa-odoo Hi :) 

I did the change and answered

About testing I don't think it's necessary, but if you really want why not :)

Cheers

**Comment by awa-odoo** [comment]
@robodoo r+ rebase-merge
